assets/src/edit-story/components/popup/index.js

```diff
@@ -19,58 +19,168 @@
  */
 import { createPortal } from 'react-dom';
 import styled from 'styled-components';
-import { useLayoutEffect, useState, useRef } from 'react';
-
-const DEFAULT_WIDTH = 270;
-const MAX_HEIGHT = 370;
+import {
+  useLayoutEffect,
+  useCallback,
+  useState,
+  useRef,
+  useEffect,
+} from 'react';
 
 const Container = styled.div.attrs(({ x, y }) => ({
-  style: { right: `${x}px`, top: `${y}px` },
+  style: { left: `${x}px`, top: `${y}px` },
 }))`
   position: fixed;
   z-index: 2147483646;
-  width: ${({ width }) => width}px;
-  max-height: ${MAX_HEIGHT}px;
-  overflow: auto;
+  ${({ placement }) => getTransforms(placement)}
 `;
 
-function Popup({ anchor, children, width = DEFAULT_WIDTH, isOpen }) {
+function getXTransforms(placement) {
+  // left & right
+  if (placement.startsWith('left')) {
+    return `-100%`;
+  } else if (placement.startsWith('right')) {
+    return null;
+  }
+  // top & bottom
+  if (placement.endsWith('-start')) {
+    return null;
+  } else if (placement.endsWith('-end')) {
+    return `-100%`;
+  }
+  return `-50%`;
+}
+
+function getYTransforms(placement) {
+  if (
+    placement.startsWith('top') ||
+    placement === 'right-end' ||
+    placement === 'left-end'
+  ) {
+    return `-100%`;
+  }
+  if (placement === 'right' || placement === 'left') {
+    return `-50%`;
+  }
+  return null;
+}
+
+// note that we cannot use percentage values for transforms, which
+// do not work correctly for rotated elements
+function getTransforms(placement) {
+  const xTransforms = getXTransforms(placement);
+  const yTransforms = getYTransforms(placement);
+  if (!xTransforms && !yTransforms) {
+    return '';
+  }
+  return `transform: ${xTransforms ? `translateX(${xTransforms})` : ``} ${
+    yTransforms ? `translateY(${yTransforms})` : ``
+  };`;
+}
+
+function getXOffset(placement, spacing = 0, anchorRect, bodyRect) {
+  switch (placement) {
+    case 'bottom-start':
+    case 'top-start':
+    case 'left':
+    case 'left-start':
+    case 'left-end':
+      return bodyRect.left + anchorRect.left - spacing;
+    case 'bottom-end':
+    case 'top-end':
+    case 'right':
+    case 'right-start':
+    case 'right-end':
+      return bodyRect.left + anchorRect.left + anchorRect.width + spacing;
```

Comment:
Should probably ensure that the popup wouldn't be out of the screen from the left-right side, too?

Reply:
Ah, looks like it might be resolved below already. _Will test in a bit_